crates/runc/src/lib.rs

```diff
@@ -0,0 +1,980 @@
+    }
+
+    pub fn debug(&mut self, debug: bool) -> &mut Self {
+        self.0.debug(debug);
+        self
+    }
+
+    pub fn log<P>(&mut self, log: P) -> &mut Self
+    where
+        P: AsRef<Path>,
+    {
+        self.0.log(log);
+        self
+    }
+
+    pub fn log_format(&mut self, log_format: LogFormat) -> &mut Self {
+        self.0.log_format(log_format);
+        self
+    }
+
+    pub fn log_format_json(&mut self) -> &mut Self {
+        self.0.log_format_json();
+        self
+    }
+
+    pub fn log_format_text(&mut self) -> &mut Self {
+        self.0.log_format_text();
+        self
+    }
+
+    pub fn systemd_cgroup(&mut self, systemd_cgroup: bool) -> &mut Self {
+        self.0.systemd_cgroup(systemd_cgroup);
+        self
+    }
+
+    // FIXME: criu is not supported now
+    // pub fn criu(mut self, criu: bool) -> Self {
+    //     self.0.criu(criu);
+    //     self
+    // }
+
+    pub fn rootless(&mut self, rootless: bool) -> &mut Self {
+        self.0.rootless(rootless);
+        self
+    }
+
+    pub fn set_pgid(&mut self, set_pgid: bool) -> &mut Self {
+        self.0.set_pgid(set_pgid);
+        self
+    }
+
+    pub fn rootless_auto(&mut self) -> &mut Self {
+        self.0.rootless_auto();
+        self
+    }
+
+    pub fn timeout(&mut self, millis: u64) -> &mut Self {
+        self.0.timeout(millis);
+        self
+    }
+
+    pub fn build(&mut self) -> Result<Client> {
+        Ok(Client(self.0.build()?))
+    }
+
+    pub fn build_async(&mut self) -> Result<AsyncClient> {
+        Ok(AsyncClient(self.0.build()?))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Client(runc::Runc);
+
+impl Client {
+    /// Create a new runc client from the supplied configuration
+    pub fn from_config(mut config: Config) -> Result<Self> {
+        config.build()
+    }
+
+    #[cfg(target_os = "linux")]
+    pub fn command(&self, args: &[String]) -> Result<std::process::Command> {
+        let args = [&self.0.args()?, args].concat();
+        let mut cmd = std::process::Command::new(&self.0.command);
+        cmd.args(&args).env_remove("NOTIFY_SOCKET"); // NOTIFY_SOCKET introduces a special behavior in runc but should only be set if invoked from systemd
+        Ok(cmd)
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    pub fn command(&self, _args: &[String]) -> Result<std::process::Command> {
+        Err(Error::Unimplemented("command".to_string()))
+    }
+
+    pub fn checkpoint(&self) -> Result<()> {
+        Err(Error::Unimplemented("checkpoint".to_string()))
+    }
+
+    fn launch(&self, mut cmd: std::process::Command, combined_output: bool) -> Result<Response> {
+        let child = cmd.spawn().map_err(Error::ProcessSpawnFailed)?;
+        let pid = child.id();
+        let result = child.wait_with_output().map_err(Error::InvalidCommand)?;
+        let status = result.status;
+        let stdout = String::from_utf8(result.stdout).unwrap();
+        let stderr = String::from_utf8(result.stderr).unwrap();
```

Comment:
Wouldn't it fail if the command outputs some non UTF-8 bytes (e.g. gzip)?

Reply:
Sure. However which is preferable in such case, returning parse error or return the raw value to caller?
(This is just my personal opinion but I think returning error is better. Higher level wrappers which use this crate, like shim, will bend the stdout/err of command to IO that they prepare in advance. And they can handle such non-utf8 within their procedure)